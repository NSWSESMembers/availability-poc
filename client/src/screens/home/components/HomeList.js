import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FlatList, Text } from 'react-native';

import { Center, Container } from '../../../components/Container';
import { Progress } from '../../../components/Progress';
import HomeEventListItem from './HomeEventListItem';
import HomeScheduleListItem from './HomeScheduleListItem';

class HomeList extends Component {
  onPressEvent = (event) => {
    const { homeNavigation } = this.props;
    homeNavigation.push('EventDetail', { eventId: event.id });
  };

  onPressNewEvent = (event) => {
    const { modalNavigation } = this.props;
    modalNavigation.push('EventNewResponse', { eventId: event.id });
  };

  onPressSchedule = (schedule) => {
    const { homeNavigation } = this.props;
    const { id, name } = schedule;
    homeNavigation.push('SchedulesDetail', { id, title: name });
  };

  getItemList() {
    const { events, schedules } = this.props.user;
    const items = [];

    const oneEvent = [];
    if (events.length > 0) {
      oneEvent.push(events[Math.floor(Math.random() * events.length)]);
    }

    // here we create 2 entries for every event to show the difference between new events and
    // ones the user has already responded to
    events.forEach((e) => {
      items.push({
        type: 'event',
        id: e.id,
        event: e,
        sortKey: 1,
      });
    });
    oneEvent.forEach((e) => {
      items.push({
        type: 'new-event',
        id: e.id,
        event: e,
        sortKey: 0,
      });
    });
    schedules.forEach((s) => {
      items.push({
        type: 'schedule',
        id: s.id,
        schedule: s,
        sortKey: s.startTime,
      });
    });

    items.sort(i => i.sortKey);

    return items;
  }

  renderItem = ({ item }) => {
    if (item.type === 'event') {
      return <HomeEventListItem event={item.event} onPress={this.onPressEvent} />;
    }
    if (item.type === 'new-event') {
      return <HomeEventListItem event={item.event} onPress={this.onPressNewEvent} urgent />;
    }
    if (item.type === 'schedule') {
      return <HomeScheduleListItem schedule={item.schedule} onPress={this.onPressSchedule} />;
    }
    throw Error(`Invalid type: ${item.type}`);
  };

  render() {
    const { loading, user, refreshing } = this.props;

    if (loading || !user) {
      return (
        <Container>
          <Progress />
        </Container>
      );
    }
    const items = this.getItemList(user);

    if (items.length === 0) {
      return (
        <Container>
          <Center>
            <Text>There is nothing interesting happening right now.</Text>
          </Center>
        </Container>
      );
    }

    return (
      <Container>
        <FlatList
          data={items}
          keyExtractor={item => `${item.type}-${item.id}`}
          renderItem={this.renderItem}
          refreshing={refreshing}
          onRefresh={this.props.onRefresh}
        />
      </Container>
    );
  }
}

HomeList.propTypes = {
  loading: PropTypes.bool,
  onRefresh: PropTypes.func.isRequired,
  refreshing: PropTypes.bool,
  modalNavigation: PropTypes.shape({
    push: PropTypes.func,
  }),
  homeNavigation: PropTypes.shape({
    push: PropTypes.func,
  }),
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    username: PropTypes.string.isRequired,
    schedules: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        details: PropTypes.string.isRequired,
        startTime: PropTypes.number.isRequired,
        endTime: PropTypes.number.isRequired,
      }),
    ),
    events: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
      }),
    ),
  }),
};

export default HomeList;
