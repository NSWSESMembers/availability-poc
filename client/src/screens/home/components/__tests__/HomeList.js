import React from 'react';
import TestRenderer from 'react-test-renderer';

import HomeList from '../HomeList';

const user = {
  id: 1,
  username: 'username',
  schedules: [
    {
      id: 1,
      name: 'schedule name',
      details: 'schedule details',
      startTime: 1524460714,
      endTime: 1524460714 + (60 * 60 * 24),
    },
  ],
  events: [
    {
      id: 1,
      name: 'event name',
      details: 'event details',
    },
  ],
};

test('renders correctly', () => {
  const tree1 = TestRenderer.create(
    <HomeList
      onRefresh={jest.fn()}
      modalNavigation={{ push: jest.fn() }}
      homeNavigation={{ push: jest.fn() }}
      user={user}
      refreshing={false}
    />,
  ).toJSON();
  expect(tree1).toMatchSnapshot();
});

test('onPressEvent', () => {
  const event = user.events[0];
  const homeNavigationPush = jest.fn();

  const renderer = TestRenderer.create(
    <HomeList
      onRefresh={jest.fn()}
      modalNavigation={{ push: jest.fn() }}
      homeNavigation={{ push: homeNavigationPush }}
      user={user}
      refreshing={false}
    />,
  );

  renderer.getInstance().onPressEvent(event);

  expect(homeNavigationPush.mock.calls.length).toBe(1);
  expect(homeNavigationPush.mock.calls[0][1].eventId).toEqual(event.id);
});

test('onPressNewEvent', () => {
  const event = user.events[0];
  const modalNavigationPush = jest.fn();

  const renderer = TestRenderer.create(
    <HomeList
      onRefresh={jest.fn()}
      modalNavigation={{ push: modalNavigationPush }}
      homeNavigation={{ push: jest.fn() }}
      user={user}
      refreshing={false}
    />,
  );

  renderer.getInstance().onPressNewEvent(event);

  expect(modalNavigationPush.mock.calls.length).toBe(1);
  expect(modalNavigationPush.mock.calls[0][1].eventId).toEqual(event.id);
});

test('onPressSchedule', () => {
  const schedule = user.schedules[0];
  const homeNavigationPush = jest.fn();

  const renderer = TestRenderer.create(
    <HomeList
      onRefresh={jest.fn()}
      modalNavigation={{ push: jest.fn() }}
      homeNavigation={{ push: homeNavigationPush }}
      user={user}
      refreshing={false}
    />,
  );

  renderer.getInstance().onPressSchedule(schedule);

  expect(homeNavigationPush.mock.calls.length).toBe(1);
  expect(homeNavigationPush.mock.calls[0][1].id).toEqual(schedule.id);
});
