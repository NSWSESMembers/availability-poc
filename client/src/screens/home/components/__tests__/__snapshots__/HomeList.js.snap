// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#EEE",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "event": Object {
            "details": "event details",
            "id": 1,
            "name": "event name",
          },
          "id": 1,
          "sortKey": 0,
          "type": "new-event",
        },
        Object {
          "id": 1,
          "schedule": Object {
            "details": "schedule details",
            "endTime": 1524547114,
            "id": 1,
            "name": "schedule name",
            "startTime": 1524460714,
          },
          "sortKey": 1524460714,
          "type": "schedule",
        },
        Object {
          "event": Object {
            "details": "event details",
            "id": 1,
            "name": "event name",
          },
          "id": 1,
          "sortKey": 1,
          "type": "event",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    invertStickyHeaders={undefined}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        progressViewOffset={undefined}
        refreshing={false}
      />
    }
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 2,
                  "elevation": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": 0,
                  "marginLeft": 10,
                  "marginRight": 10,
                  "marginTop": 6,
                  "paddingBottom": 10,
                  "paddingLeft": 8,
                  "paddingRight": 8,
                  "paddingTop": 10,
                },
                Object {
                  "backgroundColor": "#FFDDDD",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={2}
                style={
                  Object {
                    "flex": 0,
                    "fontWeight": "bold",
                    "includeFontPadding": false,
                    "marginRight": 8,
                    "textAlignVertical": "top",
                  }
                }
              >
                event name
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "color": "#6B6B6B",
                    "flex": 0,
                    "fontSize": 12,
                    "includeFontPadding": false,
                    "textAlignVertical": "bottom",
                  }
                }
              >
                event details
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "paddingLeft": 8,
                  "width": 40,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 30,
                    },
                    Object {
                      "alignSelf": "center",
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 2,
                  "elevation": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": 0,
                  "marginLeft": 10,
                  "marginRight": 10,
                  "marginTop": 6,
                  "paddingBottom": 10,
                  "paddingLeft": 8,
                  "paddingRight": 8,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "flex": 0,
                    "fontWeight": "bold",
                    "includeFontPadding": false,
                    "marginRight": 8,
                    "textAlignVertical": "top",
                  }
                }
              >
                schedule name
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={2}
                style={
                  Object {
                    "color": "#6B6B6B",
                    "flex": 0,
                    "fontSize": 12,
                    "includeFontPadding": false,
                    "textAlignVertical": "bottom",
                  }
                }
              >
                2018-04-23 to 2018-04-24
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "paddingLeft": 8,
                  "width": 40,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 30,
                    },
                    Object {
                      "alignSelf": "center",
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 2,
                  "elevation": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": 0,
                  "marginLeft": 10,
                  "marginRight": 10,
                  "marginTop": 6,
                  "paddingBottom": 10,
                  "paddingLeft": 8,
                  "paddingRight": 8,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "flex": 0,
                    "fontWeight": "bold",
                    "includeFontPadding": false,
                    "marginRight": 8,
                    "textAlignVertical": "top",
                  }
                }
              >
                event name
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={2}
                style={
                  Object {
                    "color": "#6B6B6B",
                    "flex": 0,
                    "fontSize": 12,
                    "includeFontPadding": false,
                    "textAlignVertical": "bottom",
                  }
                }
              >
                event details
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "paddingLeft": 8,
                  "width": 40,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 30,
                    },
                    Object {
                      "alignSelf": "center",
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
